docs/design/manipulation/architecture.md

```diff
@@ -0,0 +1 @@
+(TODO)
```

Comment:
Content migrated from `analysis.md`

```suggestion
## Design considerations

### Implementation in O3DE

### Use of MoveIt 2

[MoveIt 2](https://moveit.picknik.ai/humble/index.html) is the main motion
planner framework used by ROS 2. There are many ways of interacting with it. In
this design, the `ROS2ManipulationRequester` component should act as the
interface for accepting inputs in the form of request for manipulation and call
MoveIt 2 ([see the code needed to implement a Pose for the end
effector](https://moveit.picknik.ai/humble/doc/tutorials/your_first_project/your_first_project.html))
or any other motion planner.

The `ROS2ManipulationRequester` component should also take care of controlling
the status of the request and provide internal O3DE Gem information on the
status of the execution.

### Use of controllers: ros2_control and native O3DE

The integration of the [ros2_control](https://github.com/ros-controls/ros2_control) into the
architecture have some difficulties:

- controllers are typically not modular in a sense that they handle everything
 (parameters, communication and computations).
 - we would like to handle some of these elements separately for more
   flexibility and better UX.
- some of their design is not as flexible (likely since they have been only
 used with Gazebo so far).
 - for example, we would like to handle parameters in the Editor, without
   running publishers/subscribers already.

The benefits are important though. It is valuable to reuse implementations
which are tried and tested, and benefit from the joint effort of the community.
There should be the option of using native O3DE controllers if there is a good
reason to do so like big performance improvements or much better integration in
the UI.


## Architecture Proposal 
Diagram of input/outpus and data flows using ROS 2:

![Manipulation diagram](manipulation.svg)

To implement these set of actions, the O3DE simulator should have at least:

 * The capacity of publish information about the state of the joints (**RO2JointStatePub** component):
   Include the needed information from O3DE simulation related to the joints of
   the simulated robot in `sensor_msgs/JointState.msg` and publish it in the `/joint_states`
   topic. This is required to interact with ROS 2 packages (such as MoveIt)
   See [gazebo_ros_pkgs plugin for reference](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/galactic/gazebo_plugins/include/gazebo_plugins/gazebo_ros_joint_state_publisher.hpp#L44-L73)

 * Receive orders to move the different joints (**ROS2PoseTrajectory** component):
   Set the trajectory of points to be followed by joints in the O3DE simulation using as input
   `trajectory_msgs/msg/JointTrajectory.msg`
   See [gazebo_ros_pkgs plugin for reference](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/galactic/gazebo_plugins/include/gazebo_plugins/gazebo_ros_joint_pose_trajectory.hpp#L26-L48)

 * The ability to request picks or motion movements to external manipulation planners (**ROS2ManipulationRequester**):
   Request a motion plan to external planners given different types of inputs,
   a target pose for end effector or a joint state target. MoveIt `move_group_interface` interaction
   is an example of this.
   See [MoveIt 2 joint state target goal](https://github.com/ros-planning/moveit2/blob/main/moveit_ros/planning_interface/move_group_interface/include/moveit/move_group_interface/move_group_interface.h#L305-L320)
   or [MoveIt 2](https://github.com/ros-planning/moveit2/blob/main/moveit_ros/planning_interface/move_group_interface/include/moveit/move_group_interface/move_group_interface.h#L640-L554)

 * Provide data from the simulated sensors: not in the scope of this document
 
```